README.md
# DeepLearning_Projects
Deep learning projects
# Summary Task
# QnA Task
Hit 127.0.0.1:8080


news.txt
In order to restrict users from retweeting any tweet before opening the article and reading it, Twitter had started a prompt feature back in June. The support handle of Twitter had tweeted saying, “Sharing an article can spark conversation, so you may want to read it before you Tweet it. To help promote informed discussion, we're testing a new prompt on Android ― when you Retweet an article that you haven't opened on Twitter, we may ask if you'd like to open it first.” This prompt would ask a user to be sure if they would wish to share the article before reading it on the platform.

QnA.py
#!/usr/bin/env python
# coding: utf-8

# In[15]:


from transformers import pipeline
from http.server import BaseHTTPRequestHandler, HTTPServer
import json


# In[2]:


question_and_answer = pipeline("question-answering")


# In[27]:


hostName = "127.0.0.1"
serverPort = 3128


# In[8]:


mcontext = """You never saw such a commotion in all your life as when my Uncle Podger undertook to do a job. A picture would be waiting to be put up and Uncle Podger would say: ‘Oh, you leave that to me. Don’t you worry about that? I’ll do all that. Now you go and get me my hammer. You bring me the rule Tom, and I shall want the step – ladder and I had better have a kitchen chair too and Jim, you run round to Mr. Goggles and tell him Pa’s kind regards and hopes his leg’s better and will he lend him his spirit level? And don’t you go, Martha, because I shall want somebody to hold me the light, and when the girl comes back, she must go out again for a bit of picture – cord and Tom you come here I shall want you to hand me up the picture.’ And then he would lift up the picture and drop it and it would come out of the frame, and he would try to save the glass and cut himself and then he would spring round the room, looking for his handkerchief. He could not find his handkerchief, because it was in the pocket of the coat he had taken off, and he did not know where he had put the coat. The entire house had to leave off looking for his tools and start looking for his coat, while he would dance round and hinder them. ‘Doesn’t anybody in the whole house knows where my coat is? Six of you! You can’t find a coat!’ Then he’d get up and find that he had been sitting on it, and would call out. ‘Oh, you can give it up! I’ve found it myself now.’ When half an hour had been spent in tying up his finger, and a new glass had been got, and the tools, and the ladder, and the chair, and the candle had been brought, he would have another go, the whole family, including the girl and the charwoman, standing round in a semi-circle, ready to help. Two people would have to hold him there, and a fourth would hand him a nail, and a fifth would pass him up the hammer, and he would take hold of the nail, and drop it. “There!’ he would say, in an injured tone, ‘now the nail’s gone’. We would all have to go down on our knees and grovel for it, while he would stand on the chair, and grunt, and want to know if he was to be kept there all the evening. The nail would be found at last, but by that time he would have lost the hammer. “Where’s the hammer? Seven of you gaping round there and you don’t know what I did with the hammer?’ We would find the hammer and then he would have lost sight of the mark he had made on the wall. Each of us had to get up on the chair beside him and see if we could find it, and we would each discover it in a different place and he would call us all fools. And he would take the rule, and re-measure and find that he wanted half thirty – one and three-eighths inches from the corner, and would try to do it in his head, and go mad. And we would all try to do it in our heads and all arrive at different results, and the original number would be forgotten, and Uncle Podger would have to measure it again. He would use a bit of string this time and at the critical moment, when he was leaning over the chair at an angle of forty – five, and trying to reach a point on the wall, the string would slip, and down he would slide on the piano, a really fine musical effect being produced by the suddenness with which his head and body struck all the notes at the same time. At last, Uncle Podger would get the spot fixed again and put the point of the nail on it with his left hand and take the hammer in his right hand. And, with the first blow, he would smash his thumb, and drop the hammer with a yell, on somebody’s toes. Aunt Maria would observe that the next time Uncle Podger was going to hammer a nail into the wall;. she would go and spend a week with her mother while it was being done. ‘You women you make such a fuss over everything,’ Uncle Podger would reply. “I like doing a little job of this sort.’"""


# In[9]:


mquestion="""Give two words to describe Uncle Podger."""


# In[32]:


def QnA_function(context,question):
    result = question_and_answer(question=question, context=context)
    print("Answer:", result['answer'])
    print("Score:", result['score'])
    return result['answer']


# In[18]:


QnA_function(mcontext,mquestion)


# In[33]:


class MyServer(BaseHTTPRequestHandler):
    def do_POST(self):
        print("in POST") 
        content_length = int(self.headers['Content-Length']) # <--- Gets the size of data
        post_data = self.rfile.read(content_length)
        post_data_de = post_data.decode("utf-8")
        print(post_data_de)
        data = json.loads(post_data_de)
        m_context = data["context"]
        m_question = data["question"]
        answer = QnA_function(m_context,m_question)
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(bytes("<html><head><title>https://pythonbasics.org</title></head>", "utf-8"))
            #self.wfile.write(bytes("<p>Request: %s</p>" % self.path, "utf-8"))
        self.wfile.write(bytes("<body>", "utf-8"))
        self.wfile.write(bytes("<p>" + answer + "</p>", "utf-8"))
        self.wfile.write(bytes("</body></html>", "utf-8"))
        


# In[ ]:


if __name__ =="__main__":
    webServer = HTTPServer((hostName, serverPort), MyServer)
    print("Server started http://%s:%s" % (hostName, serverPort))

    try:
        webServer.serve_forever()
    except KeyboardInterrupt:
        pass

    webServer.server_close()
    print("Server stopped.")


# In[ ]:






summary.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


from transformers import pipeline
from http.server import BaseHTTPRequestHandler, HTTPServer
import time


# In[ ]:

#summarizer pipeline
summarizer = pipeline("summarization", model="t5-base", tokenizer="t5-base", framework="tf")
hostName = "127.0.0.1"
serverPort = 12345


# In[ ]:

#function to convert passage to summary
def recieve_text_to_summarize(passage):
    word_count = len(passage.split())
    summary = (summarizer(passage, min_length=5, max_length=word_count, do_sample=False))
    return summary[0]["summary_text"]


# In[ ]:

#map for cmd string to function pointer
cmd_to_function_map = {
    "summarize": recieve_text_to_summarize 
}


# In[ ]:

#http server, catering the microservice request
class MyServer(BaseHTTPRequestHandler):
#handking the GET request
    def do_GET(self):
        print("in get")
        summary = cmd_to_function_map["summarize"]("""The promise of deep learning is to discover rich, hierarchical models that represent probability""")
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(bytes("<html><head><title>https://pythonbasics.org</title></head>", "utf-8"))
        #self.wfile.write(bytes("<p>Request: %s</p>" % self.path, "utf-8"))
        self.wfile.write(bytes("<body>", "utf-8"))
        self.wfile.write(bytes("<p>" + summary + "</p>", "utf-8"))
        self.wfile.write(bytes("</body></html>", "utf-8"))
#handling the POST request  
    def do_POST(self):
        print("in post")
        content_length = int(self.headers['Content-Length']) # <--- Gets the size of data
        post_data = self.rfile.read(content_length) # reading the data
        post_data_de = post_data.decode("utf-8") #convert string in byte form to str form
        print(post_data_de)
        #if post_data_de.count("summarize")>0:
        print(post_data_de[0:9])
        if post_data_de[0:9] == "summarize": #check if summarizer string is present
            print("inside summarize")
            summary = cmd_to_function_map["summarize"](post_data_de[12:]) #nlp function call
            print(summary)
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(bytes("<html><head><title>https://pythonbasics.org</title></head>", "utf-8"))
            #self.wfile.write(bytes("<p>Request: %s</p>" % self.path, "utf-8"))
            self.wfile.write(bytes("<body>", "utf-8"))
            self.wfile.write(bytes("<p>" + summary + "</p>", "utf-8"))
            self.wfile.write(bytes("</body></html>", "utf-8"))
        else:
            print("summarize not presnt")
        #print(post_data)


# In[ ]:
#main method

if __name__ == "__main__":        
    webServer = HTTPServer((hostName, serverPort), MyServer)
    print("Server started http://%s:%s" % (hostName, serverPort))

    try:
        webServer.serve_forever()
    except KeyboardInterrupt:
        pass

    webServer.server_close()
    print("Server stopped.")


# In[ ]:






first_node_form.js
var http = require('http');
var formidable = require('formidable');
var fs = require('fs');
const axios = require('axios');
const { StringDecoder }  = require('string_decoder');


//create a server using http module

http.createServer(function (req, res) {
	var summary_data = '';
	var answer = '';
	const decoder = new StringDecoder('utf8');
	//form file upload concept
  if ((req.url == '/' || req.url =='/QnA')&& req.method.toLowerCase()=="post") {
	  console.log("/fileupload");
    var form = new formidable.IncomingForm();
	//form parser
    form.parse(req, function (err, fields, files) {
		//var oldpath = files.filetoupload.path;
		
		//res.writeHead(200, { 'Content-Type': 'application/json' });
      //res.end(JSON.stringify({ fields, files }, null, 2));
	 var radio_input = fields.input_type;
	  //console.log(radio_input);
	  //console.log("hello world");
	  //console.log(decoder.write(data));
	  //console.log(data);
		//opening and reading the uploaded file
		/*fs.readFile(oldpath, 'utf8' , (err, data) => {
		  if (err) {
			console.error(err)
			return
		  }
		  */
		  //post request to nlp server
		if(fields.input_type=="summary")
		{

			var data =  fields.description;

			console.log(data);
			  axios
				  .post('http://127.0.0.1:12345', "summarize : " + data)
				  .then(res => {
					console.log(`statusCode: ${res.status}`)
					console.log(res.data)
					//data recieved from nlp endpoint
					summary_data = res.data
				  })
				  .catch(error => {
					console.error(error)
				  })
			
			// res.write('File uploaded');
			//as nodejs is async, it calls the nlp endpoint and move to res.end(), because of it, we were not able to post the the summary data on the web page, hence to sync it, we are waiting till the summary data is not empty
			var _flagCheck = setInterval(function() {
				if (summary_data !== '') {
					clearInterval(_flagCheck);
					//ressponse to web client
					res.writeHead(200, {'Content-Type': 'text/html'});
					res.write(summary_data); // the function to run once all flags are true
					res.end();
				}
			}, 100); 
		}// interval set at 100 milliseconds

		else if(fields.input_type=="QnA")
		{
			console.log("in QnA handling");
			var context = fields.description;
			var question = fields.question_input;
			var obj = {"context": context, "question": question};
			var obj_str = JSON.stringify(obj);
			axios
				 .post("http://127.0.0.1:3128",obj_str)
				 .then(res => {
					 answer = res.data;
					 console.log(answer);
				 })
				 .catch(error =>{
					 console.log(error);
				 })
				 
		var flagcheck = setInterval(function(){
				if(answer !== '')
				{
					console.log("answer recieved");
					clearInterval(flagcheck);
					//ressponse to web client
					res.writeHead(200, {'Content-Type': 'text/html'});
					res.write(answer); // the function to run once all flags are true
					res.end();
				}
			},100);	 
		}

      
    });
  } 
  //html form, uploading txt file and post it to server
  else if(req.url=="/QnA" && req.method.toLowerCase()=="get")
  {
	 console.log("In QnA server");
	 fs.readFile("QnA.html",function(error,pgResp)
	 {
		 res.writeHead(200,{'Content-Type':'text/html'});
		 res.write(pgResp);
		 return res.end();
	 })
  }
  else {
	  console.log("/ request");
    //res.writeHead(200, {'Content-Type': 'text/html'});
	fs.readFile("index.html", function (error, pgResp) {
		res.writeHead(200, { 'Content-Type': 'text/html' });
		res.write(pgResp);
		return res.end();
	});
   /* res.write('<form action="fileupload" method="post" enctype="multipart/form-data">');
    res.write('<input type="file" name="filetoupload"><br>');
    res.write('<input type="submit">');
    res.write('</form>');
    res.end();*/
  }
}).listen(8080); 